cedar-policy-validator/src/entity_manifest.rs

```diff
@@ -0,0 +1,1103 @@
+//! Entity Manifest definition and static analysis.
+
+/*
+ * Copyright Cedar Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use std::collections::HashMap;
+use std::fmt::{Display, Formatter};
+
+use cedar_policy_core::ast::{
+    BinaryOp, EntityUID, Expr, ExprKind, Literal, PolicyID, PolicySet, RequestType, UnaryOp, Var,
+};
+use cedar_policy_core::entities::err::EntitiesError;
+use cedar_policy_core::impl_diagnostic_from_source_loc_field;
+use cedar_policy_core::parser::Loc;
+use miette::Diagnostic;
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+use smol_str::SmolStr;
+use thiserror::Error;
+
+use crate::ValidationError;
+use crate::{
+    typecheck::{PolicyCheck, Typechecker},
+    types::{EntityRecordKind, Type},
+    ValidationMode, ValidatorSchema,
+};
+
+/// Data structure storing what data is needed
+/// based on the the [`RequestType`].
+/// For each request type, the [`EntityManifest`] stores
+/// a [`RootAccessTrie`] of data to retrieve.
+#[doc = include_str!("../../cedar-policy/experimental_warning.md")]
+#[serde_as]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct EntityManifest<T = ()>
+where
+    T: Clone,
+{
+    /// A map from request types to [`RootAccessTrie`]s.
+    #[serde_as(as = "Vec<(_, _)>")]
+    #[serde(bound(deserialize = "T: Default"))]
+    pub per_action: HashMap<RequestType, RootAccessTrie<T>>,
+}
+
+/// A map of data fields to [`AccessTrie`]s.
+/// The keys to this map form the edges in the access trie,
+/// pointing to sub-tries.
+#[doc = include_str!("../../cedar-policy/experimental_warning.md")]
+pub type Fields<T> = HashMap<SmolStr, Box<AccessTrie<T>>>;
```

Comment:
Do we need the `Box`?

Reply:
Unfortunately yes, since `AccessTrie` contains a `Fields`